pepper-apis/src/main/java/org/broadinstitute/ddp/cache/CacheService.java

```diff
@@ -0,0 +1,153 @@
+        }
+    }
+
+    private <K, V> Cache<K, V> _getOrCreateCache(String cacheName, Duration entryDuration, Object mapper,
+                                                 ModelChangeType evictionModelChangeType,
+                                                 Object cacheParentObject) {
+        Cache existingCache = cacheManager.getCache(cacheName);
+        if (existingCache == null || existingCache.isClosed()) {
+            synchronized (cacheParentObject) {
+                Cache<K, V> cache = _getOrCreateCache(cacheName, entryDuration);
+                if (mapper instanceof IdToCacheKeyCollectionMapper) {
+                    cacheNameToCacheKeyCollectionMapper.put(cacheName, (IdToCacheKeyCollectionMapper) mapper);
+                } else if (mapper instanceof IdToCacheKeyMapper) {
+                    cacheNameToCacheKeyMapper.put(cacheName, (IdToCacheKeyMapper) mapper);
+                } else {
+                    throw new DDPException("Unrecognized mapper type");
+                }
+                updateChangeTypeToCacheName(evictionModelChangeType, cacheName);
+                return cache;
+            }
+        } else {
+            return (Cache<K, V>) existingCache;
+        }
+    }
+
+    private <K, V> Cache<K, V> _getOrCreateCache(String cacheName, Duration entryDuration) {
+        Cache cache = cacheManager.getCache(cacheName);
+        if (cache == null || cache.isClosed()) {
+            var cacheConfig = new MutableConfiguration<K, V>()
+                    .setExpiryPolicyFactory(CreatedExpiryPolicy.factoryOf(entryDuration))
+                    .setStatisticsEnabled(true);
+            cache = cacheManager.createCache(cacheName, cacheConfig);
+        }
+        return (Cache<K, V>) cache;
+    }
+
+    public void destroyCache(String name) {
+        cacheManager.destroyCache(name);
+    }
+
+    private void updateChangeTypeToCacheName(ModelChangeType evictionModelChangeType, String cacheName) {
+        Collection<String> existingCacheNames = modelChangeTypeToCacheName.get(evictionModelChangeType);
+        List<String> newListOfCacheNames;
+        if (existingCacheNames == null) {
+            newListOfCacheNames = List.of(cacheName);
+        } else {
+            newListOfCacheNames = new ArrayList<>(existingCacheNames);
+            newListOfCacheNames.add(cacheName);
+        }
+        modelChangeTypeToCacheName.put(evictionModelChangeType, newListOfCacheNames);
+    }
+
+    public void modelUpdated(ModelChangeType changeType, Handle handle, long id) {
+        findCacheNameByEventType(changeType).forEach(cacheName -> {
+            Cache cache = cacheManager.getCache(cacheName);
+            if (cache != null) {
+                IdToCacheKeyMapper<?> mapper = cacheNameToCacheKeyMapper.get(cacheName);
+                if (mapper != null) {
+                    Object cacheKey = cacheNameToCacheKeyMapper.get(cacheName).mapToKey(id, handle);
```

Comment:
Doesn't look like these mappers are used for anything other than removing from cache. Looks like we can simplify this and just have the plural version of the mapper. It's easy to convert the singular one to plural with `Set.of()`.

Reply:
I am expecting that the mapping to one key will be used much more commonly used than the one to the collection. Trying to keep the code needed by users of the service is minimal. But this not written in stone!